rpc/types/README.md

```diff
@@ -64,6 +64,20 @@ These mixed types are:
 
 TODO: we need to figure out a type that (de)serializes correctly, `String` errors with `serde_json`
 
+# Fixed byte containers
+TODO
+
+<!--
+
+Some fields within requests/responses are containers, but fixed in size.
+
+For example, [`crate::json::GetBlockTemplateResponse::prev_hash`] is always a 32-byte hash.
+
+In these cases, stack allocated types like `cuprate_fixed_bytes::StrArray`
+will be used instead of a more typical [`String`] for optimization reasons.
+
+-->
+
```

Comment:
@Boog900 re: https://github.com/Cuprate/cuprate/pull/147#discussion_r1654912900.

Do you want to make a `cuprate_fixed_bytes::StrArray(Bytes)` type? Or `Str<const N: usize>([u8; N])`?

If the latter, there's plenty of stack string crates, or we can do a straight copy+paste of [my impl](https://docs.rs/readable/latest/readable/str/struct.Str.html) (or a sub-set of it).

I'm also not sure the overhead of `Bytes` will be worth it in some of the types that are mostly create -> throw away.

Reply:
`Bytes` will not be efficient for parsing from json, it's only efficient is the underlying data is stored in a `Bytes` object and if the parser uses the `Bytes` functions.

So `StrArray` could be useful for epee, however in the p2p code I just use `Bytes` for strings, they aren't common.

I would want a stack string though, I would prefer using a crate, unless there is a good reason for using a homegrown one?

Reply:
> I would prefer using a crate, unless there is a good reason for using a homegrown one?

1. IIRC there's no stack string crate that everyone uses, the closest I think would be [heapless](https://github.com/rust-embedded/heapless)
1. We'd be exposing a 3rd party crate in our API, not the worst but a bit awkward IMO
2. `epee_encoding` would stop working, we'd have to make a wrapper for impl

We could also enable features to swap certain types, e.g.:
```rust
cfg_if! {
    if #[cfg(feature = "heapless")] else {
        pub type String = heapless::String;
    } else {
        pub type String = String
    }
}
```
We could do this for `ByteArray` stuff too, i.e. just keep all types as lenient as possible but allow type optimizations via features.

Reply:
Fair looking at `heapless` it does seem to be doing a similar thing to your homegrown one, and it wouldn't be too much to maintain.

Are stings common in epee binary types for RPC? I would just use `Bytes` for bin types.